### ISSUE TITLE
Mount Google Drive File System to WSL

### ISSUE THREAD

**Issue by rjh336** [microsoft/WSL#2999]
@bitcrazed I plan to follow your recommended [instructions](https://askubuntu.com/questions/9135/how-to-backup-settings-and-list-of-installed-packages) to backup `/home/<user>` and installed packages from #2955.

However, I would like to compress the backup in a tar.gz archive per [these instructions](https://help.ubuntu.com/community/BackupYourSystem/TAR), and I would like to store that archive in my Google Drive File System (GDFS). My GDFS is located in my windows file system in the G: drive. However, when I try to view this drive, I am met with the following:
```
rjh336@LAPTOP-ETTS75F8:/mnt/g$ ls /mnt
c  d  g
rjh336@LAPTOP-ETTS75F8:/mnt/g$ cd /mnt/g
rjh336@LAPTOP-ETTS75F8:/mnt/g$ ls -la
ls: reading directory '.': Function not implemented
total 0
```

Upon trying to save an archive in that location, I see a similar error message:
```
rjh336@LAPTOP-ETTS75F8:~$ tar -cvpzf /mnt/g/backup.tar.gz --exclude=./test_src/exclude_folder ./test_src/
tar (child): /mnt/g/backup.tar.gz: Cannot open: Function not implemented
tar (child): Error is not recoverable: exiting now
./test_src/
./test_src/.folder_hidden/
./test_src/.folder_hidden/.hid_file
./test_src/.hidden
./test_src/file.txt
./test_src/folder_visible/
```

Is there a proper way to mount the Google File System from windows? Or perhaps a way to mount this drive directly on WSL? I checked the File system on G: and it says it is **"FAT32"**. Any advice is greatly appreciated!

**Comment by therealkenc** [comment]
Variation on #2896 (VirtualBox additions shared folder) #2163 (WebDAV) #1962 (Windows NFS and/or Dokany) #579 (EXT2FSD). WSL DrvFS can only grok certain Windows filesystems presently.

**Comment by deven** [comment]
I currently have Windows 10 build 16299.  I installed Google Drive File Stream, and my Google Drive is visible as I: under Windows.  When I launch a WSL bash shell, this is getting automatically mounted as "/mnt/i".  While I also get the "Function not implemented" error, I have found that I can nevertheless change into subdirectories and read/write files via DrvFS to Google Drive, I can even do "ls -l" on a specific file without getting any errors.

I just can't read directory entries, which obviously makes the filesystem a little difficult to use!

Running strace shows that getdents() is the system call which is failing:

openat(AT_FDCWD, ".", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 3
getdents(3, 0x21f6890, 32768)           = -1 ENOSYS (Function not implemented)

If just this syscall could be fixed for the Google Drive File Stream filesystem, it might work perfectly.

**Comment by taylorpaul** [comment]
I am encountering the same issue!

**Comment by jvallery** [comment]
Any update on this?  This would be killer if I could work with the files on my G Drive via WSL.

**Comment by bastiao** [comment]
This will be really nice to have!

**Comment by edgarasf123** [comment]
I found that if you share the drive over the network, you can mount the share from different computer and run commands within drive folders without problems. But I was unable to mount it using localhost, it knows that you're trying to mount the local drive and just mounts it as regular drive.

![](https://i.imgur.com/Igax7eW.png)

**Comment by chris2crawford** [comment]
Here is a way to mount GFS in WSL based on [https://superuser.com/questions/1353169/getting-sshfs-working-on-wsl-or-finding-an-alternative](url)   The trick is to use [https://www.nsoftware.com/sftp/netdrive/](url) to ssh to GFS from Windows and convert in into a filesystem that can be mounted under WSL.
1) Install OpenSSH Server under Win10
    Win10 Settings -> Apps ->  Manage optional features -> Add feature -> OpenSSH Server
    restart if necessary
    open Services -> OpenSSH -> Properties -> Startup Type -> Automatic (delayed)
2) Install and run SFTPNetDrive, right click on icon in hidden icons -> Main window  (or maybe already open) -> Profile -> new profile (  server: localhost, user: crawf pwd:*** Drive Letter: F (or another)
      Advanced -> ( Protocol -> uncheck compression;  Specified folder: G:\ )  )
3) in Bash (wsl):  sudo mount -t drvfs F: /mnt/h
I'm not sure how stable it is, but I was able to open files in WSL.

**Comment by edgarasf123** [comment]
I switched to [RaiDrive](https://www.raidrive.com), haven't gotten any problems yet. And upload speeds are way better than using Google File Stream.

**Comment by riyadm** [comment]
@edgarasf123 that worked for me, thanks! 

1. Install RaiDrive, map your GDrive to a network drive.
2. In the explorer, go to Network, you should find the network location named `RaiDrive-username`.
3.  In Bash, create a new directory and [mount the network location](https://blogs.msdn.microsoft.com/wsl/2017/04/18/file-system-improvements-to-the-windows-subsystem-for-linux/): 
`$ sudo mkdir /mnt/mydrive`
`$ sudo mount -t drvfs '\\RaiDrive-username\' /mnt/mydrive`

**Comment by chris2crawford** [comment]
from: Riyad Muradov
1.	Install RaiDrive, map your GDrive to a network drive.
2.	In the explorer, go to Network, you should find the network location named RaiDrive-username.
3.	In Bash, create a new directory and mount the network location:
$ sudo mkdir /mnt/mydrive
$ sudo mount -t drvfs '\\RaiDrive-username\' /mnt/mydrive

I agree the performance is much better than SFTPNetDrive.  I got the following error on the last command:  
> mount: /mnt/mydrive: wrong fs type, bad option, bad superblock on \\RaiDrive-crawf\, missing codepage or helper program, or other error

However, just using the windows drive worked fine:
$ sudo mount -t drvfs Z: /mnt/mydrive

**Comment by katherinelauck** [comment]
Just fixed the same exact thing on my own Win10/WSL Ubuntu with @chris2crawford 's solution. I would prefer to access GFS directly because of its ability to cache files. Has anyone feature requested this to GFS?

**Comment by dos-gruut** [comment]
Hi guys, I have a simple way to do this, even this way is not actually to mount the drive. But basically, you can read or write on it.

I just wanted to copy a file from WSL into the Google Drive file stream. (For example, G: \ My drive \ hobby)
So, I wrote a `copy-to-drive.sh` which contains just one line code below.
```
cmd.exe --% /c copy 'C:\Users\dos\workspace\hobby\sample.txt' 'g:\My Drive\hobby'
```
Yes, you can invoke the windows cmd on the WSL. 
I hope this can help you guys.

**Comment by therealkenc** [comment]
> Yes, you can invoke the windows cmd on the WSL.

With recent WSL you should be able to go the other direction too under powershell with something like:

```
$ powershell.exe -Command "copy C:\\some\\path\\to\\file \\\\wsl$\\Ubuntu-18.04\\some\\directory"
```

Another (similar in spirit) way to go would be set up `ssh.exe` [on Windows](https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_overview) and use `scp.exe`, or even go the other direction and set a `sshd` server on Windows. Whatever it takes so long as the file is being read/written on the _Windows side_ and not going through WSL `drvfs`.

It is also possible to use some nested IFS ideas as suggested [earlier](https://github.com/microsoft/WSL/issues/2999#issuecomment-466703351). Nothing wrong with that in principle, but going that way is stacking one (at least potential) path to failure upon another. If it "works" so much the better, but just be aware there are a lot of unknown/untested/unsupported moving pieces involved going that route. Going through interop is at least in theory a solution that should work by design.

**Comment by Masterxilo** [comment]
I came here because of this very disappointing answer: https://askubuntu.com/questions/1094020/how-can-i-access-my-google-drive-or-external-hard-from-windows-bash

**Comment by steinbergnet** [comment]
This issue is now more than 1.5 years old.
Any updates here? Will Microsoft do anything about this?
It should not be soo hard to make 9P work with network drives.
 
Practically this is a complete showstopper for me to use WSL (1 or 2).
ALL my important files live inside GDFS. Without access to GDFS
WSL is next to useless. 

I tried using RaiDrive but the performance was much too slow for any real work.

Sharing GDFS via SMB also did not work.

Did anyone find a working solution?

**Comment by bitcrazed** [comment]
@steinbergnet and @Masterxilo - As per [@SvenGroot's response above](https://github.com/Microsoft/WSL/issues/2999#issuecomment-386465812), this issue is likely caused by GDrive misbehaving. You might want to file an issue on Google.

I advise this reluctantly - I hate it too when vendors seem to point fingers - but WSL just treats drives as ... well ... drives. If the drive behaves, one can access the drive from WSL:

![image](https://user-images.githubusercontent.com/961950/67110973-ce453880-f188-11e9-93ae-f384d5277244.png)

And I don't want to sound biased, but if GDrive isn't able/willing to make this change, you might perhaps have to consider other cloud file sync providers.

**Comment by steinbergnet** [comment]
From what I see it looks like you are accessing a local directory on
your C: drive that might be mirrored to OneDrive, but not OneDrive itself.
I tried accessing my OneDrive (mounted as separate drive) and it
failed the same way as GDFS.

**Comment by therealkenc** [comment]
> If the drive behaves, one can access the drive from WSL:

That's "not wrong". These third-party IFS problems _can_ be solved on the Windows driver side. Here's hoping that happens. There is a credible counter line of reasoning though. The IFS isn't misbehaving on Windows, and working on Windows is a pretty good definition of working.

In the case of Google Drive, a constructive path forward would be to try to convince them to officially [support](https://itsfoss.com/use-google-drive-linux/) Linux. Or for someone highly motivated to do it for them. If one took [this](https://github.com/odeke-em/drive) and [this](https://github.com/hanwen/go-fuse) one could probably get there with _effort_.

> Sharing GDFS via SMB also did not work.

Made me look. Works here. Leads credibility to thesis it ought work on 9p, becuase same thing modulo implementation detail.

![image](https://user-images.githubusercontent.com/7049700/67116691-d2c41e00-f195-11e9-9b1f-92942057209a.png)

**Comment by judge2020** [comment]
@therealkenc That's the wrong Google Drive system. Drive FIle Stream is for GSuite accounts and opening files pulls from the server on-demand, instead of downloading everything and syncing.

https://www.google.com/drive/download/

![image](https://user-images.githubusercontent.com/8601934/67117123-e2506080-f1af-11e9-96c0-7b09311dfaf7.png)


[wsl 1]

**Comment by therealkenc** [comment]
> Drive FIle Stream is for GSuite accounts

Ah. Don't have one of those. So you're faced with the double whammy of no one without a GSuite account able to repro let alone suggest a fix. Most unfortunate.

**Comment by Rondom** [comment]
Given that GSuite is a paid for product, I think it does not do any harm to open a support ticket with them.

**Comment by steinbergnet** [comment]
Let's be realistic. Google Drive is not misbehaving. It works as designed
as a drive
under Windows. The fact that WSL2 somehow does not handle it correctly can
hardly be considered a bug, especially since WSL2 is still in development.
Clearly the problem lies somewhere in the WSL2/9P code.
Since Microsoft is propagating WSL it should have a look at this issue.

**Comment by SvenGroot** [comment]
WSL as a file system client is a bit more strict about how a file system needs to behave than Windows in general. This goes both for WSL 1 and WSL 2, because while the mechanism of actually getting to the files is quite different, the actual code that interacts with the file system is mostly the same. For example, we require file systems to support one of a few directory information classes that we use to enumerate directories. Windows has more information classes than the ones that WSL supports, but they don't provide the information WSL requires.

DrvFs is the component of WSL that interacts with native Windows file systems. In this context, that also refers to the 9p server used to access Windows files on WSL 2. DrvFs was originally designed to access only NTFS. We've long since updated it to support the various flavors of FAT supported by Windows, CDFS, and SMB (with the main target being a Windows SMB server sharing an NTFS volume). Everything that works besides that is pretty much just lucky, not officially supported.

I would definitely not mind taking a look at GDrive to see what we do that mystifies it, and if it's something we can work around, but that's pretty much only possible if I can get it under a debugger. To do that, I need access to a machine from someone who as an account, since it's a paid service so I don't have one.

**Comment by chris2crawford** [comment]
Dear Sven,
  I really appreciate you taking a look.  I have had some success with
RaiDrive, but its performance lags behind GFS, and occasionally I get the
blue screen, for example, when starting up emacs from the shared drive.
  Email me at crawford@pa.uky.edu to arrange for remote access to my
account.
--Chris

**Comment by therealkenc** [comment]
Thanks for the details Sven; your write-ups are always appreciated.

> Windows has more information classes than the ones that WSL supports, but they don't provide the information WSL **requires**.

I have been half-meaning to ask for some time (but didn't) whether it is possible/practical to loosen our definition of "requires" for a few (probably common) pain points. For example I took a [run once](https://github.com/microsoft/WSL/issues/579#issuecomment-399668778) at getting [Dockany](https://github.com/dokan-dev/dokany) mounted on WSL. Didn't work; they don't implement [`FILE_STANDARD_LINK_INFORMATION`](https://processhacker.sourceforge.io/doc/ntioapi_8h_source.html#l00484). I didn't bother you with it at the time; not your problem. It is fixable in Dockany for someone motivated, totally.

But this said, can we reasonably or realistically  _survive_ without (in this arbitrary example) `NumberOfAccessibleLinks`? If we fail to get `FILE_STANDARD_LINK_INFORMATION`, fall back to less esoteric `FILE_STANDARD_INFORMATION`, which you _gotta know_ the filesystem implements, or they'd fall down on Windows. If WSL "requires" the schmancy `NumberOfAccessibleLinks` (to track deleted links), punt on hard links altogether. Hard links are [not well supported](https://github.com/libfuse/sshfs/issues/109) on some FUSE filesystems anyway; for example, the canonically popular [sshfs](https://github.com/libfuse/sshfs). So in the case of missing required info (and again we're just using Dockany as an illustrative example here) say: "[no hardlinks for you](https://images.ctfassets.net/23wh7e5ryr18/photo-48559/da360458ad8b8593c3d3cce7d9524367/48559-seinfeld-soup-nazi.jpg?fm=jpg&fl=progressive&w=900&q=50)". `link(2)` fails by-design and `stat.st_nlink` for files is 1. [I bet a nickle GDrive doesn't support NTFS hard links either, even if that is almost certainly not the blocker here.]

> we require file systems to support one of a few directory information classes that we use to enumerate directories

Understood. You don't say which ones, but, similar to the link information example, nothing too schmancy (high `_FILE_INFORMATION_CLASS` identifier) can near by definition be a hard _requirement_ of WSL, for the fact that you've been able to SMB2 mount a Windows filesystem on Real Linux for decades, before all the esoteric info classes were invented. Which is why I asked the (still pending) follow up regarding (quoth) "Sharing GDFS via SMB also did not work". If GDrive doesn't support enough `*INFORMATION` classes to SMB share a directory, that would be (a) a little surprising, and (b) plainly a limitation of that filesystem. [I'm at a disadvantage to either confirm or deny said limitation without GDrive.]

**Comment by bitcrazed** [comment]
Superb write-up as usual @SvenGroot, thank you! :)

I'll work with @craigloewen-msft to see if we can get a GDRive account & app to debug against.

**Comment by phaabe** [comment]
Do you still work on that issue? Will WSL 2 solve this?

**Comment by Rondom** [comment]
Google Drive Stream does use Dokany.
One can use the Dokan mirror example to track it down without a Google Business account.

https://support.google.com/drive/answer/1716931
https://github.com/google/google-drive-dokany

**Comment by therealkenc** [comment]
> Google Drive Stream does use Dokany.

😮

Uh, how the hell did I miss this fact in when I was Googling the topic last year. In this case patching Dockany is a viable path forward. The only reason I didn't at the time was laziness; or being overly generous, was putting my time sink allotment towards other things.

**Comment by query1000** [comment]
I encountered the same issue and my workaround is that I use "google backup and sync".
It syncs the files under C: so it is naturally mounted on WSL.
I don't know the difference between 'backup and sync' and 'drive file stream' though :-D

**Comment by judge2020** [comment]
@query1000 it has access to team drives and whatnot, it's also only for G Suite accounts. https://support.google.com/drive/answer/7638428?hl=en

**Comment by jminne** [comment]
G Drive filestream also only downloads files on demand so you don't have to
specify folders in include or exclude from your sync.

Thanks for the workaround suggestion.

John Minne

**Comment by JeffreyShran** [comment]
I am assuming that this is the same reason that I'm unable to execute Python scripts from mounted Google file systems too?
```
jeff@DESKTOP-T2NRS6G:/mnt/g/My Drive/Hack_The_Planet/scripts$ ./jeffPy.py
Traceback (most recent call last):
  File "./jeffPy.py", line 3, in <module>
    import requests
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 963, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 906, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1280, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1252, in _get_spec
  File "<frozen importlib._bootstrap_external>", line 1368, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1408, in _fill_cache
OSError: [Errno 38] Function not implemented: '/mnt/g/My Drive/Hack_The_Planet/scripts'
```

The script runs fine on the same system in a non Google mount.

I'm afraid the technicalities into why `ls` and various things don't work are beyond my scope of knowledge. However if there is anything we can do to offer the intelligent folks here some support please shout and if I have the skills, I will.

**Comment by braderhart** [comment]
> Here is a way to mount GFS in WSL based on [https://superuser.com/questions/1353169/getting-sshfs-working-on-wsl-or-finding-an-alternative](url) The trick is to use [https://www.nsoftware.com/sftp/netdrive/](url) to ssh to GFS from Windows and convert it to a filesystem that can be mounted under WSL.
> 
> 1. Install OpenSSH Server under Win10
>    Win10 Settings -> Apps ->  Manage optional features -> Add feature -> OpenSSH Server
>    restart if necessary
>    open Services -> OpenSSH -> Properties -> Startup Type -> Automatic (delayed)
> 2. Install and run SFTPNetDrive, right click on icon in hidden icons -> Main window  (or maybe already open) -> Profile -> new profile (  server: localhost, user: *** pwd:*** Drive Letter: F (or another)
>    Advanced -> ( Protocol -> uncheck compression;  Specified folder: G:\ )  )
> 3. in Bash (wsl):  sudo mkdir /mnt/f; sudo mount -t drvfs F: /mnt/f
>    I'm not sure how stable it is, but I was able to open files in WSL.

This isn't necessary in WSL 2. You can just use sshfs from WSL to mount your Google Drive from Windows.

**Comment by jerivas** [comment]
I've had great success with [google-drive-ocamlfuse](https://github.com/astrada/google-drive-ocamlfuse) to connect my Drive directly to WSL (not using a Windows client at all). The only hiccup is that during the initial Authorization you will need to copy the oAuth URL manually (as WSL can't open the browser in Windows with the `open` command). It's just a matter of creating a shim for `xdg-open` to print the URL in the terminal instead: https://github.com/astrada/google-drive-ocamlfuse/wiki/Headless-Usage-&-Authorization#simple-another-way

This also works well with Shared Drives (previously Team Drives) for G Suite customers: https://github.com/astrada/google-drive-ocamlfuse/wiki/Team-Drives.

**Comment by scivision** [comment]
Now that WSL2 is officially released, wondering if any WSL2 users have had luck using Google Drive Stream directly (without the workarounds above)

**Comment by judge2020** [comment]
Unfortunately, no:

![image](https://user-images.githubusercontent.com/8601934/85169725-4abd7900-b23a-11ea-8118-b18c241a96dc.png)

**Comment by braderhart** [comment]
@scivision As long as you install OpenSSH for Windows, and sshfs in WSL, then you can mount it that way without any third-party software.

**Comment by scivision** [comment]
Okay thank you braderhart I missed that final comment you made in your comment a month ago. sorry for not reading more closely and thanks for responding.

**Comment by hans-t** [comment]
@braderhart could you elaborate what's the exact command of sshfs to mount? Thank you.

**Comment by braderhart** [comment]
@hans-t 

1. Install OpenSSH for Windows:
https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_install_firstuse
2. Install `sshfs` within your WSL 2 Linux distro (e.g. `sudo apt-get install sshfs`)
3. Edit `/etc/fuse.conf` and uncomment `user_allow_other` or run the command bellow from your WSL 2 terminal:
    ```
    sudo sed --in-place '/user_allow_other/s/^#//g' /etc/fuse.conf
    ```
4. Create a mount path:
    ```
    sudo mkdir /mnt/drive \
    && sudo chown `id --user`:`id --group` /mnt/drive
    ```
5. Copy the following to a file called `sshfs_drive.sh` and give it execute permissions (`chmod +x sshfs_drive.sh`). Make sure to edit `<username>` and `<identity_file>` with your own (e.g. `~/.ssh/id_ed25519`):
    ```
    #!/bin/sh
    
    DRIVE_PATH='G:/My Drive'
    HOST_IP=`tail --lines=1 /etc/resolv.conf | awk '{print $2}'`
    IDENTITY_FILE='<identity_file>'
    USERNAME='<username>'
    
    sshfs -o "IdentityFile=$IDENTITY_FILE" $USERNAME@$HOST_IP:"$DRIVE_PATH" -o allow_other,gid=`id --group`,idmap=user,uid=`id --user` /mnt/drive
    ```

**Comment by gabriellovate** [comment]
@braderhart I didn't clearly understand which information IDENTIFY_FILE and USERNAME should receive. are those variables related to unix files or windows files?

**Comment by braderhart** [comment]
@gabriellovate Those are related to the the Windows SSH server you setup initially.

**Comment by YutaItoh** [comment]
@gabriellovate 
You might also want to refer to this article:
https://askme4tech.com/how-install-and-configure-open-ssh-server-windows-10
To make `ssh` from WSL to Windows work, I had to tweak the permission of the `%UserProfile%/.ssh/authorized_keys` in a Windows manner.

@braderhart 
Great post! You saved my day.

One last behavior of `sshfs` that I haven't been able to resolve is that it mounts the root of GDFS even if I set a subfolder like:
`DRIVE_PATH='G:/My Drive/my_desired_subfolder'`

**Comment by braderhart** [comment]
@YutaItoh Glad I was able to help. I am not able to replicate the issue you mention of mounting the root of GDFS when using a subfolder. I just tried and it correctly mounted the subfolder.

**Comment by YutaItoh** [comment]
@braderhart oh, please never mind that comment. I just had a typo in my modified script. Your solution works like a charm. thx!

**Comment by zemega** [comment]
> 
> 
> @hans-t
> 
>     1. Install OpenSSH for Windows:
>        https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_install_firstuse
> 
>     2. Install `sshfs` within your WSL 2 Linux distro (e.g. `sudo apt-get install sshfs`)
> 
>     3. Edit `/etc/fuse.conf` and uncomment `user_allow_other` or run the command bellow from your WSL 2 terminal:
>        ```
>        sudo sed --in-place '/user_allow_other/s/^#//g' /etc/fuse.conf
>        ```
> 
>     4. Create a mount path (e.g. `/mnt/drive`):
>        ```
>        sudo mkdir /mnt/drive \
>        && sudo chown `id --user`:`id --group` /mnt/drive
>        ```
> 
>     5. Copy the following to a file called `sshfs_drive.sh` and give it execute permissions (`chmod +x sshfs_drive.sh`). Make sure to edit `<username>` and `<identity_file>` with your own (e.g. `~/.ssh/id_ed25519`):
>        ```
>        #!/bin/sh
>        
>        DRIVE_PATH='G:/My Drive'
>        HOST_IP=`tail --lines=1 /etc/resolv.conf | awk '{print $2}'`
>        IDENTITY_FILE='<identity_file>'
>        USERNAME='<username>'
>        
>        sshfs -o "IdentityFile=$IDENTITY_FILE" $USERNAME@$HOST_IP:"$DRIVE_PATH" -o allow_other,gid=`id --group`,idmap=user,uid=`id --user` /mnt/drive
>        ```
> 
> 
> > Note: If you don't want to use a script, then you can copy the command from the script and modify it accordingly

This works with your own GDFS account and folders. This however doesn't work when the a folder is shared by someone else. For example the shared folder path in Windows wouuld appear as  below:
`G:\.shortcut-targets-by-id\15unoo6ctOXQDcJjGbz8T1ePfAXrYoVZZ\Atmosperic Group`
Any idea how to access shared folder in GDFS?

**Comment by braderhart** [comment]
@zemega I just tested this out. You can modify the script and replace `DRIVE_PATH` with the shared folder path, and it works. Not sure if that is where you're encountering the error, but it works for me. The script is more of an example, but if you wanted to create multiple GDFS mounts on startup you'd need to modify the script to include separate variables and mount points, or mount the entire `G:/` and use symlinks.

I'd also suggest you checkout rclone which has more options available:

https://rclone.org/commands/rclone_mount/

**Comment by zemega** [comment]
> 
> 
> @zemega I just tested this out. You can modify the script and replace `DRIVE_PATH` with the shared folder path, and it works. Not sure if that is where you're encountering the error, but it works for me. The script is more of an example, but if you wanted to create multiple GDFS mounts on startup you'd need to modify the script to include separate variables and mount points, or mount the entire `G:/` and use symlinks.
> 
> I'd also suggest you checkout rclone which has more options available:
> 
> https://rclone.org/commands/rclone_mount/

I test it again. It does work after all. I'm not sure what I did wrong previously.

**Comment by reidrankin** [comment]
On WSL2, this appears to be caused by dokan-dev/dokany#932, and should be fixed by dokan-dev/dokany#933.

**Comment by timothychen1999** [comment]
Not working, it cannot even get the directory.

**Comment by braderhart** [comment]
@timothychen1999 I only have limited bandwidth to help out. What is it that you're trying to do? I see some of the folders listed there. Can you paste the script, output from the `ls -la`, etc in a gist or pastebin and I'll take a look?

**Comment by timothychen1999** [comment]
@braderhart 
Basicly, I run the script from you,
![image](https://user-images.githubusercontent.com/45344933/96541060-c9eab880-12d1-11eb-9479-6541a23af315.png)
Script I used: https://gist.github.com/timothychen1999/051f8f6968f926b7f2a303915d203d6e
The only difference I can imagine is that I am using a Zh-TW system, so instead of the folder "My drive", the dir is called "我的雲端硬碟".
The ls -la command in mnt/drive gives
`drwx------ 1 timothy timothy 0 Oct 20 12:36 '$RECYCLE.BIN'
drwx------ 1 timothy timothy 0 Oct 20 12:36  .file-revisions-by-id
drwx------ 1 timothy timothy 0 Oct 20 12:36  .shortcut-targets-by-id`

The ultimate target is to get vscode in wsl to interact with my gdrive.
Sorry for incomplete information.

**Comment by koru1130** [comment]
@timothychen1999 
Have the same problem, zh-TW system too.
I tried to use soft link (junction) to avoid the "共用雲端硬碟" and it works.

**Comment by braderhart** [comment]
@koru1130 That is great news. I'm not able to utilize another language to test this out, but it sounds like you may have found a workaround. Is the problem with the `My Drive` on `G:/`?

**Comment by timothychen1999** [comment]
@koru1130 Well, Can you give more details on how you achieve that?
Cause as I know junction can only work on NTFS while GoogleDrvie seems to be in FAT32

**Comment by reidrankin** [comment]
Just to update everyone, dokan-dev/dokany#932 has now been fixed. Applications using Dokan (i.e. Google Drive) will need to update their copy of dokan1.dll, but dropping in a newer copy manually *should* work.

Fixing this should avoid the need for an sshfs setup and let you mount using drvfs as usual. Please let me know if this setup exhibits the same issues with Unicode characters in paths!

**Comment by braderhart** [comment]
@reidrankin Are you pushing out a new release now? I can't find the download link anywhere.

**Comment by chris2crawford** [comment]
Thanks, Reid!  Does it also work with WSL1?
--Chris

**Comment by zemega** [comment]
@reidrankin So we have to wait for Google Drive File Stream to update on its own? Could not find dokan1.dll in the Dokan github. While in GDFS program folder there are dokan3229.dll at 0 bytes and dokancc3229.dll at 1,266 KB. From where can we download the replacement file and which file t replace?

**Comment by reidrankin** [comment]
@braderhart @zemega It's not part of a release build yet; [here's](https://ci.appveyor.com/project/Maxhy/dokany/builds/35742135) the relevant snapshot build.

@chris2crawford @zemega I don't have access to a Google Drive File Stream account, so I appear to have made some incorrect assumptions -- namely, that the app was using Dokan the way any other app would, through `dokan1.dll`. Apparently, Google is special, so they've got their own [fork](https://github.com/google/google-drive-dokany) which uses a C++ wrapper layer they've named `dokancc`. They've also squashed the git history so it's difficult to tell what they've changed or what commit it was forked *from*. The version number in `dokan.h` suggests it was pulled from v1.0.0 (in 2017) but I haven't checked any of the file hashes.

So no dice on the simple drop-in-a-dll solution, I'm afraid. I guess you'd need to try updating that library and building a new dokancc3229.dll -- or bugging Google to do it themselves.

**Comment by Philosophistoria** [comment]
@timothychen1999
I hope you doing well afterwards.
I have a similar problem with ja_JP system.
But it has successfully worked when I modified the path expression to win-like from linux-like, using backslash `\` instead of slash `/`, just like:
```
DRIVE_PATH="G:\共有ドライブ\<subdir>"  
```
Also, I found that 2-byte chars like cannot be listed with `ls`
When I newly make three directories such as `test` `テスト`, ` 試験` just under the <subdir>, I could only see `test` directory. Also the other two dir are not only unable-to-be-seen but somehow also unable-to-be-accessed.  

My Environment: Win 20H2, Ubuntu20.04

I hope this can help someone.

**Comment by timothychen1999** [comment]
@Philosophistoria Thanks for the help.
Still no luck on that though. I am still getting  No such file or directory from the sshfs.
I am on ubuntu 18.04, maybe I'd try 20.04 later.

**Comment by vvigilante** [comment]
> So no dice on the simple drop-in-a-dll solution, I'm afraid. I guess you'd need to try updating that library and building a new dokancc3229.dll -- or bugging Google to do it themselves.

No luck yet, the public source code is 18 months old, for version 1918 and lacks a fair amount of interface functions needed by the GoogleDrive executable. The older lib will not link against the newer executable.

**Comment by vvigilante** [comment]
I made a [patch to the binary](https://github.com/vvigilante/drivefs-wsl2-patch).
It is not the cleanest job in the world, but it works for me.

**Comment by Liryna** [comment]
@vvigilante Thank you for your patch!
I am the Dokan maintainer and I am also working on Google DriveFS.
I have merged https://github.com/dokan-dev/dokany/pull/933 changes in the Google dokan driver sources for the next release 👍

**Comment by zemega** [comment]
@Liryna Would it be safe to assume that starting from  47.0.17.0, Google Drive is accessible from WSL without any problem? I just noticed that `ls` works. I have not tried any full script or heavy usage of Google Drive through WSL.

**Comment by Liryna** [comment]
The @reidrankin patch was imported to fix the listing issue, after that I am not aware of any issue that was reported for WSL. If any issue is faced, please try with Dokan and send a feedback.

**Comment by zemega** [comment]
@Liryna I'm having an issue with `mv`. For example,
`touch a`
`mv a b`
`mv: cannot move 'a' to 'b': No such file or directory`

`cp` works fine. Where should I send feedback?

**Comment by Liryna** [comment]
First try with the Dokan mirror sample to see if the issue comes from Google drive or Dokan.
If it is Dokan, you can directly report on the repository.

**Comment by Liryna** [comment]
@zemega I have been able to reproduce the issue. I see where and why it is happening. I will work on a fix. Thanks for reporting the issue!

**Comment by Masterxilo** [comment]
It seems this is fixed now?!
On WSL 2, Windows 20H2 (set to German), Google Drive Version 47.0.19.0
```
windows@DESKTOP-I608RI4:/mnt/g/Meine Ablage/Data$ echo hello > hello-cr.txt
windows@DESKTOP-I608RI4:/mnt/g/Meine Ablage/Data$ ls
```

**Comment by Dr-Irv** [comment]
Here is a workaround that provides some minimal shell capabilities for GDrive.  I needed to inspect a file on Google Drive Stream using `tail`, and found that using the `bash` supplied with `Git` on windows allowed me to do that.  My Google Drive is G: in DOS:
```bash
 G:\>"c:\Program Files\Git\bin\bash

<REDACTED> MINGW64 /g
$ ls
'$RECYCLE.BIN'  'My Drive'  'Shared drives'

<REDACTED> MINGW64 /g
$ ls My\ Drive/*.JPG
'My Drive/IMG_2039.JPG'  'My Drive/IMG_2040.JPG'  'My Drive/IMG_2041.JPG'
```

**Comment by nelsonjchen** [comment]
Not sure if this matters for anyone but "Google Drive for desktop" seems to work well for my gmail Google account. Drive File Stream did not work in the past. This seems to be a new Drive File Stream that works for all Google accounts.

https://support.google.com/a/answer/7491144/deploy-google-drive-for-desktop?utm_medium=et&utm_source=aboutdrive&utm_content=getstarted&utm_campaign=en_us&hl=en

https://support.google.com/googleone/answer/10309431/upcoming-changes-to-google-drive-sync-clients?hl=en#zippy=%2Cwill-anything-happen-to-my-files-when-drive-for-desktop-releases%2Cwhy-are-you-making-this-change

In other words, a G Suite account is no longer required to test this functionality.

**Comment by traversjames** [comment]
So it looks like "Google Drive for desktop" access now works for WSL2 but [USB/serial ports don't](https://docs.microsoft.com/en-us/windows/wsl/wsl2-faq#can-i-access-the-gpu-in-wsl-2-are-there-plans-to-increase-hardware-support). Accessing these hardware resources works on WSL1 but "Google Drive for desktop" doesn't. I need both :weary:.

Since I looked at this issue last, however, it seems Google have re-enabled "Backup and Sync" - they seem to flip-flop a lot on their drive offerings. This means that the synced files are directly accessible on WSL1 with hardware port access. "Google Drive for Desktop" can be run in parallel - maybe this was obvious but this now covers all use cases for me.

**Comment by pesobreiro** [comment]
Hello everyone, following the idea of @nelsonjchen and @traversjames, if this still matters. But this is truly annoying when using google drive. I think the simple approach is to install [backup and sync](https://www.google.com/drive/download/ basically the same thing with [few functionalities](https://support.google.com/drive/answer/7638428?hl=en). Google drive download per use vs the files being available locally in backup and sync, which allows to access them using the mount of the windows drive in wsl.

**Comment by Liryna** [comment]
WSL 1 compatibility was fixed in the latest Dokan release. Google will ship this version in a coming version.

**Comment by adamsd5** [comment]
I upgraded to Google Drive for Desktop yesterday.   I cannot access /mnt/g/.
ls: cannot access '/mnt/g': Invalid argument

Using WSL2 and Ubuntu 20.04LTS

**Comment by sglaser** [comment]
The initial issue works for me (did nothing special, someone must have fixed things).

Unfortunately, there's another minor issue now:

For all files seen from my Google Drive (paid version), Ubuntu under WSL sees Inode #2. This causes find(1) to complain about file system loops that aren't real. For example:

    $ find . -print
    find: File system loop detected; ‘./Alloy’ is part of the same file system loop as ‘.’.
    ...

    $ ls -l -d -i . ./Base-6.0
    2 drwxr-xr-x 0 sglaser sglaser 512 Jul 22 15:33 ./
    2 drwxr-xr-x 0 sglaser sglaser 512 Jun 29 10:43 ./Base-6.0/
    ...

**Comment by Liryna** [comment]
@sglaser this is an known issue and will be fixed in version 50.

**Comment by Liryna** [comment]
Google Drive version 50 is released that should cover all issues mentioned here.
@therealkenc I believe this can now be closed.

**Comment by madzayasodara** [comment]
Trying to mount my gdrive on WSL2 using rclone and getting this.

```
sudo rclone mount gdrivebrown ../../mnt/d
2021/09/04 21:04:03 mount helper error: fusermount: fuse device not found, try 'modprobe fuse' first
2021/09/04 21:04:03 Fatal error: failed to mount FUSE fs: fusermount: exit status 1
```
```
modprobe fuse
modprobe: ERROR: ../libkmod/libkmod.c:586 kmod_search_moddep() could not open moddep file '/lib/modules/4.4.0-19041-Microsoft/modules.dep.bin'
modprobe: FATAL: Module fuse not found in directory /lib/modules/4.4.0-19041-Microsoft
```
```
sudo apt install fuse
Reading package lists... Done
Building dependency tree
Reading state information... Done
fuse is already the newest version (2.9.4-1ubuntu3.1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] y
Setting up friendly-recovery (0.2.31ubuntu2.1) ...
/usr/sbin/grub-probe: error: failed to get canonical path of `rootfs'.
dpkg: error processing package friendly-recovery (--configure):
 subprocess installed post-installation script returned error exit status 1
Errors were encountered while processing:
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

Appreciate any help on the issue. 

```

**Comment by switchbrad** [comment]
@madzayasodara 

I don't know if a reverse mount will work like that. As long as you mount the drive from Window in rclone.exe, it should be accessible inside WSL2.

**Comment by braderhart** [comment]
@madzayasodara I don't think you can mount with your WSL2-distro rclone package. You'd need to mount with the rclone.exe in Windows and then should be able to access it inside your WSL2 distro.

**Comment by madzayasodara** [comment]
I also have my gdrive mounted as a network drive, which I have mounted using mountain duck, but cannot access it from my WSL2....

**Comment by madzayasodara** [comment]
Ok I mounted gdrive as a drive using the new Google Drive and then mounted in WSL2
`sudo mount -t drvfs D: /mnt/d`
seem to be working now

**Comment by vinhdizzo** [comment]
@Liryna I just migrated to Google Drive version 51.0.9.0 on Windows 10.  I am on WSL 2.  G drive shows up on `/mnt/g` in bash, but I still get the error:
```
/mnt/g$ ls -lrtha
ls: reading directory '.': Invalid argument
total 0
```

I can navigate to the drive with no problem:
```
/mnt/g$ ls -lrtha
ls: reading directory '.': Invalid argument
total 0
```

Am I doing something wrong?  Thanks!

**Comment by Liryna** [comment]
@vinhdizzo could you share the exact command line you used to mount the drive ?

**Comment by vinhdizzo** [comment]
@Liryna when I launch bash, the G drive is already mounted as `/mnt/g`, so I didn't need to mount.  I could unmount and remount, and I get the following:
```
$ sudo umount /mnt/g
$ sudo umount g
$ cd /mnt/g
$ ls -lrtha
ls: reading directory '.': Invalid argument
total 0
```

**Comment by Liryna** [comment]
@vinhdizzo what happens if you try to mount by using the same commande line in other answers?

**Comment by vinhdizzo** [comment]
@Liryna  Here's what I get:
```
$ sudo mount -t drvfs '\\localhost\G' /mnt/g
mount: /mnt/g: special device \\localhost\G does not exist.
$ sudo mount -t drvfs '\\192.168.1.123\G' /mnt/g
mount: /mnt/g: special device \\192.168.1.123\G does not exist.
$ sudo mount -t drvfs '\\192.168.1.123\G$' /mnt/g
mount: /mnt/g: special device \\192.168.1.123\G$ does not exist.
```

C drive is OK:
```
sudo mount -t drvfs '\\192.168.1.123\c' /mnt/g
```

That was the only command I saw in the thread that is based on Google Drive and not other third party tools.  Did you want me to try anything else?  Thanks!

**Comment by Liryna** [comment]
@vinhdizzo please give a try to:
sudo mount -t drvfs G: /mnt/g

**Comment by vinhdizzo** [comment]
@Liryna sorry I did try that first but forgot to paste it.  This is what I get:
```
$ sudo mount -t drvfs G: /mnt/g
$ cd /mnt/g
$ ls
ls: reading directory '.': Invalid argument
```

**Comment by sudeshjethoe** [comment]
this works for me, but sometimes it helps to unmount first, e.g. umount /mnt/g (as root ofcourse)

**Comment by pgwilliams** [comment]
I've no idea when this started working (it wasn't the last time I checked - perhaps 6 months ago) but today it's working fine.   Within my WSL Ubuntu I navigated to /mnt/  and there was my G drive (mapped as G:\), and from there I went to /mnt/g/My Drive/  and all my files were there as expected.   Happy Customer 😄 👍

**Comment by ilsaloving** [comment]
This only works for your personal documents.  If you want to use a Shared Drive, it is completely inaccessible.  In fact, it is unusable from anything other than explorer.  I tried several different tools in addition to WSL, and they all either displayed nothing, or just a .lnk to some inaccessible location.

**Comment by chris2crawford** [comment]
I use WSL plus shared drives using Google Drive for Desktop.  But it only works for shared drives in my own organization, not shared from other companies.  Not sure why.

On Dec 6, 2022 13:01, ilsaloving ***@***.***> wrote:
CAUTION: External Sender


This only works for your personal documents. If you want to use a Shared Drive, it is completely inaccessible. In fact, it is unusable from anything other than explorer. I tried several different tools in addition to WSL, and they all either displayed nothing, or just a .lnk to some inaccessible location.

**Comment by steinbergnet** [comment]
Hello,

I always start Windows Explorer from WSL2/bash using the mounted Google
Drive file system.

Since the last WSL Update this week I started to get this error message
after running
WSL for a few hours when I run explorer from bash:

$ explorer.exe G:
<3>WSL (2209) ERROR: UtilBindVsockAnyPort:295: bind failed 38

Rebooting will fix the issue, but it will reappear after a few hours.

Any ideas how to fix this?

Is there a way to downgrade WSL to the previous version?

Cheers
Dirk

**Comment by literakl** [comment]
I have used Debian to search files on G drive successfully but suddenly it stopped working. Tried Google drive restart without luck:

literakl@DESKTOP-590U5T0:/$ ls -l /mnt/
ls: cannot access '/mnt/g': Invalid argument
ls: cannot access '/mnt/l': Invalid argument
total 0
drwxrwxrwx 1 literakl literakl 512 Jan 14 09:11 c
d????????? ? ?        ?          ?            ? g

Edice	Windows 10 Home
Verze	22H2
Datum instalace	‎30.‎11.‎2022
Build operačního systému	19045.2486
Výkon	Windows Feature Experience Pack 120.2212.4190.0
